src/brogue/Items.c

```diff
@@ -6033,6 +6048,52 @@ void relabel(item *theItem) {
     }
 }
 
+// If the most recently equipped item caused another item to be unequiped, is
+// uncursed, and both haven't left the inventory since, swap them back.
+void swap() {
+    item *theItem;
+    char buf[COLS * 3], swappedOutName[COLS], swappedInName[COLS];
+    unsigned char command[10];
+
+    if (rogue.swappedIn == NULL || rogue.swappedOut == NULL) {
+        confirmMessages();
+        message("You have nothing to swap.", false);
+        return;
+    }
+
+    if (rogue.swappedIn->flags & ITEM_CURSED) {
+        itemName(rogue.swappedIn, swappedInName, false, false, NULL);
+        sprintf(buf, "You can't; your %s appears to be cursed.", swappedInName);
```

Comment:
This feels under-abstracted. Is there a way to have all this (incl. ring updates and strength check below) happen via the usual methods?

Reply:
Probably, though it would probably involve me touching `equip` and `unquip` as well.

For both actions, there is a "high-level" function (`equip`, `unequip`) that calls a "low-level" one (`equipItem`, `unequipItem`.)  Much of the game logic lives in the low-level functions (which seems correct to me, and is certainly convenient to us now), while the prompting, messaging, etc are handled in the high-level ones.

Except that there is just a smidge of game logic in the high-level ones as well.  Namely the ring update and strength check you called out.  I don't know why the ring update is placed in `equip` instead of `equipItem`, it doesn't seem to generate any messages or interaction; it seems to just update player properties.  The strength check is perhaps a little more understandable, it does generate messages (but no prompts) while also updating encumbrance.   Perhaps there is a need for the low-level functions to be quiet.

I will experiment with pushing the ring update and strength check from the high-level to the low-level (and in the process, remove them from the swap action here.)

The curse detection & messaging might similarly be pushed down into `unequipItem`, it does already have the logic but no messaging.  `swap` (or whatever we call it) would then be on the hook to notice that nothing happened after the `unequipItem`, but this is easily checked.

Reply:
...now that I look, both `equipItem` and `unequipItem` already call `updateRingBonuses`.  I think the `updateRingBonuses` calls in both `unequip` and `swap` can be dropped.

The strength check is a little trickier to wave away.  It (only) messages the player if needed and delegates to `updateEncumbrance`.  `updateEncumbrance` updates the player's move and attacks speeds to match their status, and then delegates to `recalculateEquipmentBonuses`.

And `equipItem` already calls `recalculateEquipmentBonuses` for the armor and weapon cases.  `unequipItem` already calls `updateEncumbrance` unconditionally.

To me, it seems safe to alter `equipItem` to call `updateEncumbrance` unconditionally, and drop the calls to `recalculateEquipmentBonuses`.  This would still keep `equipItem` silent, something else would need to call `strengthCheck` to print "can barely lift" at the appropriate times.

Alternately `equipItem` might replace its weapon & armor `recalculateEquipmentBonuses` calls with `strengthCheck` calls, and insert a new call to `updateEncumbrance` in the ring case.  Then `strengthCheck` could be dropped from `equip` and `swap`.

Incidentally I imagine the reason these lower level functions are trying to be silent is to allow them to run many times within a given player turn.  I'm willing to believe that `equipItem` and `unequipItem` are still only called the expected number of times in various scenarios, so adding messaging to them seems safe from that standpoint.

Reply:
Ok.  The game logic changes are in the earlier commits leading up to the final one, and can be reviewed individually.  

`updateRingBonuses` and `strengthCheck` have been pushed down.

`equipItem`, `unequipItem` now perform the cursed checks, and have boolean return values (true on success.)  They also now directly handle cursed messaging.

Together these changes allow `equip`, `unequip` and `swap` (now `swapLastEquipment`) to be simpler, and for a moderate amount of dead/redundant code to be removed.